File: 0980-unique-paths-iii/0980-unique-paths-iii.js
/**
 * @param {number[][]} grid
 * @return {number}
 */
var uniquePathsIII = function(grid) {
    var paths = 0 ;
    let emptySquares = 1 ; // counting destination as emptySquare
    let startRow, startCol ;
    
    for(let i=0; i<grid.length; i++){
        for(let j=0; j<grid[0].length; j++){
            if(grid[i][j] == 0){
                emptySquares++ ;
            }
            else if(grid[i][j] == 1){
                startRow = i ;
                startCol = j ;
            }
        }
    }
    function dfs(r,c,visited,target,grid){
        
        if(r<0 || c<0 || r>=grid.length || c>=grid[0].length || grid[r][c] == -1) return ;
        
        if(grid[r][c] == 2){
            if(visited == target) paths++ ;
            return ;
        }
        
        let temp = grid[r][c] ;
        grid[r][c] = -1 ; 
        
        dfs(r,c+1,visited+1,target,grid) ;
        dfs(r,c-1,visited+1,target,grid) ;
        dfs(r+1,c,visited+1,target,grid) ;
        dfs(r-1,c,visited+1,target,grid) ;
        
        grid[r][c] = temp ;
    }
    dfs(startRow,startCol,0,emptySquares,grid) ;
    return paths ;
};

File: 0980-unique-paths-iii/0980-unique-paths-iii.java
class Solution {
    int walks = 0 ;
    public int uniquePathsIII(int[][] grids) {

        int[] source = {-1,-1} ,destination = {-1,-1} ;
        int emptySquares = 0 ;

        for(int i=0; i<grids.length; i++){
            for(int j=0; j<grids[0].length; j++){
                if(grids[i][j] == -1) continue;
                else if(grids[i][j] == 1){
                    source = new int[]{i,j} ;
                    grids[i][j] = 0 ;
                }
                else if(grids[i][j] == 2){
                    destination = new int[]{i,j} ;
                    grids[i][j] = 0 ;
                }
                emptySquares++ ;
            }
        }

        f(source[0],source[1], destination, grids, 1, emptySquares) ;
        return walks ;
    }

    private void f(int r, int c, int[] dest, int[][] grids, int count, int target){

        if(r == dest[0] && c == dest[1]){
            if(count == target) walks++ ;
            return ;
        }

        grids[r][c] = -1 ;

        int[][] moves = {{0,1},{0,-1},{1,0},{-1,0}} ;

        for(int i=0; i<4; i++){
            int x = r + moves[i][0] ;
            int y = c + moves[i][1] ;

            if(isValid(x,y,grids)){
                f(x,y,dest,grids,count+1,target) ;
            }
        }

        grids[r][c] = 0 ;
    }
    private boolean isValid(int r, int c, int[][] grids){
        return r>=0 && c>=0 && r<grids.length && c<grids[0].length && grids[r][c] == 0 ;
    }
}